### PR TITLE
Make LIFO/HIFO accounting methods O(n*log(m))

### PR DESCRIPTION
**Pull request by qwhelan**
The `LIFO` and `HIFO` methods are unfortunately quadratic time (first test is trivial for `LIFO`):
```
One In, One Out
          FIFO      LIFO      HIFO
1     0.000268  0.000233  0.000233
10    0.001612  0.001335  0.001329
100   0.011682  0.012137  0.020656
1000  0.125671  0.126606  1.090273

Buys, then Sells
          FIFO      LIFO      HIFO
1     0.000235  0.000238  0.000233
10    0.001277  0.001378  0.001477
100   0.012547  0.021762  0.029432
1000  0.123024  1.113870  1.903485
```
The cases are the same as in https://github.com/eprbell/rp2/pull/115 and assume the `FIFO` improvements there are merged.

The overall approach is to filter out exhausted lots early, so they do not need to be considered in each pass. In `FIFO`, this can be done by keeping track of another index. In `LIFO` and `HIFO` we introduce a heap where "active" lots are kept. The general `for` loop structure is kept to remove exhausted lots (as may occur when switching accounting methods across years), with the first non-exhausted lot guaranteed to be the one we are searching for. The use of heaps improves each loop to `O(log m)`.

We translate the search criteria into a `heap_key` as Python only natively supports min heaps. It is likely that this key will need to be refined to guarantee deterministic behavior.

Due to test failures, this is not ready to merge but posting to start getting feedback.

And the results with this PR:
```
One In, One Out
          FIFO      LIFO      HIFO
1     0.000247  0.000250  0.000241
10    0.001286  0.001341  0.001632
100   0.011826  0.012560  0.012876
1000  0.123382  0.131119  0.133895

Buys, then Sells
          FIFO      LIFO      HIFO
1     0.000239  0.000262  0.000243
10    0.001301  0.001400  0.001477
100   0.012038  0.013003  0.015297
1000  0.125055  0.137297  0.171604
```

### CODE REVIEW

File: src/rp2/abstract_accounting_method.py
Comment (eprbell) [inline src/rp2/abstract_accounting_method.py:84]: Instead of conditional code can we capture this through inheritance? Specifically: 
* add a `_get_accounting_method_iterator` abstract method here;
* implement `_get_accounting_method_iterator` in FIFO, LIFO and HIFO plugins;
* call `_get_accounting_method_iterator` inside `__iter__` here.
* rename `AccountingMethodIterator` to `ListAccountingMethodIterator`
* add a `AbstractAccountingMethodIterator` superclass to `ListAccountingMethodIterator` and `HeapAccountingMethodIterator`
* have `__iter__` return `AbstractAccountingMethodIterator` rather than the `Union`

**Review by eprbell — CHANGES_REQUESTED** [review]
This looks really interesting. I have only skimmed it for now, but I'm providing some early feedback.

File: src/rp2/abstract_accounting_method.py
Comment (qwhelan) [reply src/rp2/abstract_accounting_method.py:84]: Thanks for the feedback, I'll take a stab over the weekend.

Comment (qwhelan) [reply src/rp2/abstract_accounting_method.py]: Updated with your suggestions - I had to add some accessors but otherwise should match what you have here.

Some tests are failing but I'm guessing that's a matter of fixing `heap_key()`/etc such that the sort is stable.

Comment (eprbell) [reply src/rp2/abstract_accounting_method.py]: Sounds good, thanks. After I merged #115 some conflicts appeared here. After they are fixed I'll take another look.

Comment (qwhelan) [reply src/rp2/abstract_accounting_method.py]: Conflicts are resolved, needed a force push to update though

Comment (eprbell) [reply src/rp2/abstract_accounting_method.py]: OK, no worries.

File: src/rp2/abstract_transaction.py
Comment (eprbell) [inline src/rp2/abstract_transaction.py:90]: This seems unused and also it exposes to the user that internal id is an int (currently this is an implementation detail). Can we remove this property?

File: src/rp2/in_transaction.py
Comment (eprbell) [inline src/rp2/in_transaction.py:210]: What is this used for?

File: src/rp2/abstract_accounting_method.py
Comment (eprbell) [inline src/rp2/abstract_accounting_method.py:57]: I'm thinking rather than having a conditional on `use_heap`, can we make this method abstract and implement it in the subclasses? If I'm correct that should allow us to:
* remove `use_heap` altogether;
* move `add_selected_lot_to_heap` out of the abstract class and into hifo and lifo as a private method (or maybe into a `AbstractHeapAccountingMethod` superclass);
* move  `heap_key` out of the abstract class and into hifo and lifo as a private method (or maybe into a `AbstractHeapAccountingMethod` superclass);
* perhaps move `ListAccountingMethodIterator` and `HeapAccountingMethodIterator` in the subclasses too.

This would have the advantage of keeping this superclass free of implementation details.

File: src/rp2/abstract_accounting_method.py
Comment (eprbell) [inline src/rp2/abstract_accounting_method.py:96]: This looks like a protected rathen than public method: it should start with `_`.

File: src/rp2/plugin/accounting_method/fifo.py
Comment (eprbell) [inline src/rp2/plugin/accounting_method/fifo.py:72]: It would be great not to have heap-related methods in FIFO (see my other comment at line 57 of abstract_accounting_method.py on this topic).

File: src/rp2/abstract_accounting_method.py
Comment (eprbell) [inline src/rp2/abstract_accounting_method.py:58]: Not sure I'm getting this loop: could you clarify the intent?

**Review by eprbell — CHANGES_REQUESTED** [review]
I get the gist of what you're trying to do (use priority queues for hifo and lifo, instead of lists) and I think it holds a lot of promise. I went through the code and I'm providing some initial feedback: feel free to push back if any of my comments doesn't make sense to you (the code is still in progress and I may have misunderstood a few things).

File: src/rp2/in_transaction.py
Comment (qwhelan) [reply src/rp2/in_transaction.py:210]: Needed to define a sort order on transactions: https://docs.python.org/3/library/functions.html#sorted

File: src/rp2/abstract_transaction.py
Comment (qwhelan) [reply src/rp2/abstract_transaction.py:90]: This isn't used in the commits I've published but it is needed to get existing tests to pass as is. For example, with `LIFO` and `test_data2.ods`:
 - Rows 6 and 7 have the same transaction time of `2020-02-01T11:18Z`
 - Existing `LIFO` logic will consume row 7, then 6
 - `heapq.pushheap()` however will have row 6 at the front of the heap
 - To maintain equivalent behavior, we need a tie-breaker such that row 7 < row 6.
 - The existing `internal_id` is a `str` and thus not ideal (as we want to negate it so `-7 < -6`)
 
Let me know if there's a better candidate for fixing `test_test_data2_rp2_full_report`

File: src/rp2/abstract_accounting_method.py
Comment (qwhelan) [reply src/rp2/abstract_accounting_method.py:50]: The float is due to `LIFO` using the negative of the timestamp for the sort key, we could potentially make that `RP2Decimal`.

As for the tuples, I'm still trying to figure out a combo of keys such that existing tests pass. Once I have those, I'll clean up with one of the options you mentioned.

File: src/rp2/abstract_accounting_method.py
Comment (qwhelan) [reply src/rp2/abstract_accounting_method.py:58]: So it's getting called here within the accounting_engine: https://github.com/eprbell/rp2/blob/main/src/rp2/accounting_engine.py#L212

The lots under consideration are determined by the timestamp of the `taxable_event` here: https://github.com/eprbell/rp2/blob/main/src/rp2/accounting_engine.py#L202-L204

This loop is simply advancing the heap to the timestamp of the `taxable_event`.

File: src/rp2/abstract_accounting_method.py
Comment (qwhelan) [reply src/rp2/abstract_accounting_method.py:63]: I've got all of your requests implemented other than:
> I'm thinking rather than having a conditional on use_heap, can we make this method abstract and implement it in the subclasses?

Just to clarify, this is currently on `LotCandidates` so no subclasses (unless you mean move all of this logic into `AbstractHeapAccountingMethod`)

> - move add_selected_lot_to_heap out of the abstract class and into hifo and lifo as a private method (or maybe into a AbstractHeapAccountingMethod superclass);

This is somewhat coupled to the above and not seeing an obvious solution. I'll think about it a bit more Sunday.

**Comment by qwhelan** [comment]
Unit tests are now passing (other than a new false alarm on an external link)

File: src/rp2/abstract_accounting_method.py
Comment (eprbell) [reply src/rp2/abstract_accounting_method.py:63]: Good point. One possible solution:
* `LotCandidates` becomes a hierarchy of classes (parallel to `AbstractAccountingMethod`)
* `AbstractAccountingMethod` gets a new `create_lot_candidates()` abstract method, which gets implemented in the subclasses
* the [instantiation of LotCandidates](https://github.com/eprbell/rp2/blob/main/src/rp2/accounting_engine.py#L122) in `AccountingEngine` becomes `node.value.create_lot_candidates()`

However before we go too far down the refactoring rabbit hole, let's get this PR to work: we can reevaluate how much we want to refactor at the end.

File: src/rp2/abstract_transaction.py
Comment (eprbell) [reply src/rp2/abstract_transaction.py:90]: Ah, interesting. I think the clean way to do this is to add a `row` property to `AbstractTransaction`, which is something I wanted to do for some time. Basically `row` is the row where the transaction appears in the spreadsheet. This row value is already passed to the transaction creation [during parsing](https://github.com/eprbell/rp2/blob/main/src/rp2/ods_parser.py#L128) and it is currently used to initialize the `internal_id` of the transaction as a string. 

So the change would be to rename the [internal_id](https://github.com/eprbell/rp2/blob/main/src/rp2/abstract_transaction.py#L33) constructor parameter of `AbstractTransaction` to `row` and then use the row value to initialize both the `internal_id` as today (still a string) and a new `row` field which would be an `int`.

Then your code could simply use the new `row` read-only property.

So we have two options, I think: we can either do the above or just rename `internal_id_int` to `row` for now and do the above change later (maybe add a TODO to the code to avoid losing track of it).

**Comment by eprbell** [comment]
> Unit tests are now passing (other than a new false alarm on an external link)

Oh, fantastic! Let me take another look.

**Comment by qwhelan** [comment]
@eprbell Refactor is complete and all tests are passing, but I'm guessing we'll need a round of cleanup or two. I didn't see your suggestion on `row` until just now but I'll implement that this week with any cleanup you suggest.

**Comment by eprbell** [comment]
> @eprbell Refactor is complete and all tests are passing, but I'm guessing we'll need a round of cleanup or two. I didn't see your suggestion on `row` until just now but I'll implement that this week with any cleanup you suggest.

Great! I'll review.

File: src/rp2/abstract_transaction.py
Comment (qwhelan) [reply src/rp2/abstract_transaction.py]: Just uploaded a commit to handle this. The semantics are weird if `row=` is not passed (defaults to `id(self)`) but it's a bit invasive to fix that, as kwarg order would have to be modified in `OutTransaction`/etc.

File: src/rp2/plugin/accounting_method/hifo.py
Comment (eprbell) [inline src/rp2/plugin/accounting_method/hifo.py:32]: Change the comment at line 42 in positive terms, similar to what we did for FIFO (i.e.: "This loop features...").

File: src/rp2/plugin/accounting_method/lifo.py
Comment (eprbell) [inline src/rp2/plugin/accounting_method/lifo.py:34]: Change the comment at line 44 in positive terms, similar to what we did for FIFO (i.e.: "This loop features...").

File: src/rp2/plugin/accounting_method/lifo.py
Comment (eprbell) [inline src/rp2/plugin/accounting_method/lifo.py:63]: Could you clarify this bit?

File: src/rp2/plugin/accounting_method/hifo.py
Comment (eprbell) [inline src/rp2/plugin/accounting_method/hifo.py:55]: Could you clarify this change?

File: src/rp2/plugin/accounting_method/hifo.py
Comment (eprbell) [inline src/rp2/plugin/accounting_method/hifo.py:62]: Could you clarify this bit (it's the same as in lifo)?

File: src/rp2/abstract_transaction.py
Comment (eprbell) [inline src/rp2/abstract_transaction.py:42]: Could you elaborate a bit on this comment? I'm not clear what "fallback case" this refers to.

**Review by eprbell — CHANGES_REQUESTED** [review]
I have a few comments and questions here and there, but this looks really good: it'll be very beneficial to people with lots of transactions. We will also have to update the [plugin developer documentation] to mention the two plugin flavors (list vs heap).

File: src/rp2/in_transaction.py
Comment (qwhelan) [reply src/rp2/in_transaction.py:210]: Done

File: src/rp2/abstract_accounting_method.py
Comment (qwhelan) [reply src/rp2/abstract_accounting_method.py:180]: Done

File: src/rp2/abstract_accounting_method.py
Comment (qwhelan) [reply src/rp2/abstract_accounting_method.py:56]: Not easily, it breaks a lot of stuff due to `_get_accounting_method_iterator()` needing to take `AbstractAcquiredLotCandidates` due to the Liskov substitution principle.

File: src/rp2/plugin/accounting_method/hifo.py
Comment (qwhelan) [reply src/rp2/plugin/accounting_method/hifo.py:55]: Mostly relevant for method changes across years - we want to skip already exhausted lots.

File: src/rp2/plugin/accounting_method/lifo.py
Comment (qwhelan) [reply src/rp2/plugin/accounting_method/lifo.py:63]: If the selected lot will not be entirely consumed by the current transaction, add it back to the top of the heap.

File: src/rp2/plugin/accounting_method/hifo.py
Comment (qwhelan) [reply src/rp2/plugin/accounting_method/hifo.py:62]: Same as below - if there's any leftover amount in the lot add it back to the heap.

File: src/rp2/abstract_accounting_method.py
Comment (eprbell) [reply src/rp2/abstract_accounting_method.py:56]: Perhaps it's easier to show what I mean [with a diff](https://github.com/user-attachments/files/15750687/heap_in_heap_classes_only.txt). This passes both static analysis and tests.

File: src/rp2/abstract_accounting_method.py
Comment (eprbell) [reply src/rp2/abstract_accounting_method.py:41]: Yes, you're right: let's leave this as is then.

File: src/rp2/abstract_accounting_method.py
Comment (eprbell) [reply src/rp2/abstract_accounting_method.py:125]: We can't do this easily, as discussed above.

File: src/rp2/abstract_accounting_method.py
Comment (eprbell) [reply src/rp2/abstract_accounting_method.py:198]: We won't do this: see comment above.

File: src/rp2/plugin/accounting_method/hifo.py
Comment (eprbell) [reply src/rp2/plugin/accounting_method/hifo.py:62]: Oh, I see, it's because you popped it earlier, so you need to push it back on the heap: this bit is necessary because it's a heap-based class (I was still thinking in terms of FIFO). Makes sense.

File: src/rp2/plugin/accounting_method/lifo.py
Comment (eprbell) [reply src/rp2/plugin/accounting_method/lifo.py:63]: Same as HIFO.

File: src/rp2/plugin/accounting_method/hifo.py
Comment (eprbell) [reply src/rp2/plugin/accounting_method/hifo.py:55]: I see, and you don't need the `selected_acquired_lot.spot_price < acquired_lot.spot_price` logic because it is implicit via heap operatons. Don't you need a similar if with the break in the LIFO case?

File: src/rp2/abstract_transaction.py
Comment (qwhelan) [reply src/rp2/abstract_transaction.py]: Updated in newest commits

File: src/rp2/plugin/accounting_method/lifo.py
Comment (qwhelan) [reply src/rp2/plugin/accounting_method/lifo.py:34]: Done in newest commits

File: src/rp2/plugin/accounting_method/hifo.py
Comment (qwhelan) [reply src/rp2/plugin/accounting_method/hifo.py:32]: Done

File: src/rp2/abstract_accounting_method.py
Comment (qwhelan) [reply src/rp2/abstract_accounting_method.py:56]: Seems I was missing the asserts, I've added your patch

File: src/rp2/abstract_accounting_method.py
Comment (qwhelan) [reply src/rp2/abstract_accounting_method.py:185]: We get an analysis error if removed:
```
 W0223: Method 'lot_candidates_order' is abstract in class 'AbstractAccountingMethod' but is not overridden in child class 'AbstractListAccountingMethod' (abstract-method)
```

File: src/rp2/plugin/accounting_method/hifo.py
Comment (qwhelan) [reply src/rp2/plugin/accounting_method/hifo.py]: This seems to be unnecessary as an equivalent check is made a few lines up. I've removed in the latest commit

File: src/rp2/abstract_accounting_method.py
Comment (eprbell) [reply src/rp2/abstract_accounting_method.py:193]: OK, let's leave it then.

**Review by eprbell — APPROVED** [review]
LGTM! There are a couple of minor things still TBD (like updating the dev docs), but we shouldn't get blocked on that (I can work on that separately). Thanks again, for this awesome set of optimizations!

**Pull request merged**
